Async handling of Onion Messages with `BackpressureQueue`

gijswijs opened the pull request:
This pull request refactors the onion message handling to be asynchronous, addressing potential performance bottlenecks in the synchronous design from #10089. This PR also depends on #9838 (commit 3511604e7b18291965cdffa0c564ccf62539b was cherry-picked into this PR). This PR is meant to be merged after those two PRs have been merged.

The main changes are:

1.  **Introduction of `BackpressureQueue`**: A new generic, bounded queue is added. This queue uses a "Random Early Drop" (RED) predicate to start randomly dropping incoming messages as it approaches capacity. This prevents the queue from blocking when full and provides backpressure to the network under heavy load.

2.  **Asynchronous Message Processing**: The `onion_message.OnionEndpoint` is modified to use this new `BackpressureQueue`.
    *   When a new onion message is received from a peer, instead of being processed immediately, it is added to the backpressure queue. This allows the peer's main read loop to continue without waiting for the onion packet to be decrypted and forwarded.
    *   The endpoint now runs its own background goroutine that dequeues and processes messages, containing the forwarding logic that was introduced in the second PR.

In summary, where the previous PRs introduced the *ability* to forward onion messages, this PR makes the implementation robust. It ensures that a high volume of onion messages won't block critical peer communication, improving the node's stability and throughput as a router in the onion message network.

Roasbeef commented on onion_message/onion_endpoint.go:
```diff
@@ -137,16 +168,76 @@ func (o *OnionEndpoint) CanHandle(msg msgmux.PeerMsg) bool {
+	}()
+
+	for {
+		result := o.onionMsgQueue.Dequeue(ctx)
```
Can be worked into the `actor` system as mentioned in other PRs.

lightninglabs-deploy commented:
@gijswijs, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@gijswijs, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@gijswijs, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@gijswijs, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@gijswijs, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@gijswijs, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@gijswijs, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@gijswijs, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@gijswijs, remember to re-request review from reviewers when ready

Roasbeef commented on onion_message/onion_endpoint.go:
```diff
@@ -108,6 +120,16 @@ type OnionEndpoint struct {
 	MsgSender OnionMessageSender
+
+	// onionMsgQueue is a queue that contains incoming onion messages.
+	onionMsgQueue *queue.BackpressureQueue[msgmux.PeerMsg]
```
Hmm, this should be used to instantiate the actor itself.

lightninglabs-deploy commented:
@gijswijs, remember to re-request review from reviewers when ready

gijswijs commented:
Rebased the PR after the merge of #10089. Also updated the description to better explain what the BackpressureQueue is doing in the context of the onion message actor.

Roasbeef commented on queue/back_pressure.go:
```diff
@@ -0,0 +1,231 @@
+	select {
+
+	case item := <-q.ch:
+		return fn.Ok(item)
```
Should still check this.

Roasbeef commented on actor/actor.go:
```diff
@@ -93,8 +103,14 @@ func NewActor[M Message, R any](cfg ActorConfig[M, R]) (*Actor[M, R],
+	// the default ChannelMailbox.
+	var mailbox Mailbox[M, R]
+	if cfg.MailboxFactory != nil {
+		mailbox = cfg.MailboxFactory(ctx, mailboxCapacity)
```
👍
